### ISSUE TITLE
Numerical instability problem

### ISSUE THREAD

**Issue by BensonRen** [sagelywizard/pytorch-mdn#6]
First want to thank the creator of this repo, although simple it is a great and convenient implementation of MDN for the community.

During training, there is numerical instability problems arising from the multiplication of the probability that can be changed to addition.

**Comment by huukim136** [comment]
[https://github.com/sagelywizard/pytorch-mdn/blob/bb8152902ae628a136e6038913d012e1583609ca/mdn/mdn.py#L78](url)
agreed. if the target's dimension is large. it will be problematic when multiplying all the probabilities. maybe we `torch.mean()` can be used instead of `torch.prod()`.

**Comment by charliebudd** [comment]
I can confirm this behaviour for a 31 dimension output, and that `torch.sum()` works (and I assume `torch.mean()` too). Perhaps a sanity check after the `torch.prod()` and fallback to the above methods if it fails.

**Comment by leeedwina430** [comment]
I can confirm this behavior even for a 5 dimension input and 1 dimension output. My solution is to use 

`from torch.distributions.normal import Normal`

to rewrite the function `mdn_loss(pi, sigma, mu, target)` with:

```
def mdn_loss(pi, sigma, mu, target):
     model_dist = Normal(mu, sigma)
     target = target.unsqueeze(1).expand_as(sigma) 
     log_prob = model_dist.log_prob(target) 
     weighted_log_prob = torch.log(pi).unsqueeze(1).expand_as(sigma) + log_prob 
     return -torch.logsumexp(weighted_log_prob, dim=1).mean()
```

Since the built-in function `.lob_prob()' and 'torch.logsumexp' is much more stable than self-implemented functions.

**Comment by charliebudd** [comment]
Yeah, I eventually arrived at the same solution...

```python

def gaussian_probability(pi, sigma, mu, target):
    """Returns the probability of `target` given MoG parameters `sigma` and `mu`.

    Arguments:
        sigma (BxGxO): The standard deviation of the Gaussians. B is the batch
            size, G is the number of Gaussians, and O is the number of
            dimensions per Gaussian.
        mu (BxGxO): The means of the Gaussians. B is the batch size, G is the
            number of Gaussians, and O is the number of dimensions per Gaussian.
        target (BxI): A batch of target. B is the batch size and I is the number of
            input dimensions.

    Returns:
        probabilities (BxG): The probability of each point in the probability
            of the distribution in the corresponding sigma/mu index.
    """
    target = target.unsqueeze(1).expand_as(sigma)
    
    # old version....
    # ret = ONEOVERSQRT2PI * torch.exp(-0.5 * ((target - mu) / sigma)**2) / sigma
    # probs_old = pi * torch.prod(ret, 2)
    
    log_ret = math.log(ONEOVERSQRT2PI) - 0.5 * ((target - mu) / sigma)**2 - torch.log(sigma)
    log_prob = torch.log(pi) + log_ret.sum(2)
    
    probs = torch.exp(log_prob)
    
    return probs


def mdn_loss(pi, sigma, mu, target):
    """Calculates the error, given the MoG parameters and the target

    The loss is the negative log likelihood of the data given the MoG
    parameters.
    """
    prob = gaussian_probability(pi, sigma, mu, target)
    nll = -torch.log(torch.sum(prob, dim=1))
    return torch.mean(nll)
```

**Comment by jbeck9** [comment]
In case anyone else comes across this issue, just another note for training stability.

For most applications where the target variable isn't a clean and defined distribution, you will most likely want to add a small epsilon value to sigma when computing gaussian_probability. SGD may result in the model trying to produce Gaussians with a variance close to zero, which will cause the loss to explode.
